wip: limit unapplied conjuncts effect on selectivity

cucaroach opened the pull request:
Potential fix, works but don't like it and I'm %100 sure there's a better fix I haven't thought of yet.

cucaroach committed:
opt: limit unapplied conjuncts effect on selectivity

Instead of counting every constraint on a set of columns limit it to 4.

Epic: none
Fixes: #100510
Release note: None

wip: opt_tester statement-bundle support

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/101004)
<!-- Reviewable:end -->

cucaroach committed:
opt: use w/ on row estimate for lookup join when cols form a key

Fixes: #100510
Epic: none
Release note (sql change): Make the optimizer less likely to pick bad
indexes when lookup joins have some indexes where the lookup columns
form a key and some that don't.
